samples/99_zil_interpreter/app/builtins.rb

```diff
@@ -37,9 +37,25 @@ def expect_minimum_argument_count!(arguments, min_count)
 ZIL_BUILTINS[:LVAL] = define_for_evaled_arguments { |arguments, context|
   expect_argument_count!(arguments, 1)
   var_atom = arguments[0]
-  raise FunctionError, "No local value for #{var_atom.inspect}" unless context.locals.key? var_atom
 
-  context.locals[var_atom]
+  result = :ZIL_LOCAL_VALUE_NOT_ASSIGNED
+  [context.locals, *context.locals_stack].each { |stack|
+    if stack.key? var_atom
+      result = stack[var_atom]
+      break
+    end
+  }
+
+  raise FunctionError, "No local value for #{var_atom.inspect}" if result == :ZIL_LOCAL_VALUE_NOT_ASSIGNED
+  result
+}
+
+ZIL_BUILTINS[:GVAL] = define_for_evaled_arguments { |arguments, context|
+  expect_argument_count!(arguments, 1)
+  var_atom = arguments[0]
+  raise FunctionError, "No local value for #{var_atom.inspect}" unless context.globals.key? var_atom
```

Comment:
someone might want to change `local` to `global` here 😉
possibly @kfischer-okarin with his godly world-altering abilities?

Reply:
Cleaned up in new commit. 👍